File: lib/screens/edit_service_screen.dart
import 'package:flutter/material.dart';

import '../models.dart';
import '../theme.dart';
import '../widgets/ui.dart';

class EditServiceScreen extends StatefulWidget {
  final Vehicle vehicle;
  final String? serviceId;
  final VoidCallback onBack;
  final ValueChanged<Vehicle> onUpdateVehicle;

  const EditServiceScreen({super.key, required this.vehicle, this.serviceId, required this.onBack, required this.onUpdateVehicle});

  @override
  State<EditServiceScreen> createState() => _EditServiceScreenState();
}

class _EditServiceScreenState extends State<EditServiceScreen> {
  ServicePackage? _existing;
  late String _name;
  late Set<String> _selected;

  @override
  void initState() {
    super.initState();
    _existing = widget.serviceId != null
        ? getServiceRecords(widget.vehicle).where((s) => s.id == widget.serviceId).firstOrNull
        : null;
    _name = _existing?.name ?? '';
    _selected = {...(_existing?.categoryIds ?? const [])};
  }

  void _toggle(String id) {
    setState(() {
      if (_selected.contains(id)) {
        _selected.remove(id);
      } else {
        _selected.add(id);
      }
    });
  }

  void _save() {
    final trimmed = _name.trim();
    if (trimmed.isEmpty) return;
    final categoryIds = widget.vehicle.categories.map((c) => c.id).where(_selected.contains).toList();
    if (categoryIds.isEmpty) return;
    widget.onUpdateVehicle(upsertService(widget.vehicle, ServicePackage(id: _existing?.id, name: trimmed, categoryIds: categoryIds)));
    widget.onBack();
  }

  Future<void> _remove() async {
    final existing = _existing;
    if (existing == null) return;
    final confirmed = await showDialog<bool>(
      context: context,
      builder: (ctx) => AlertDialog(
        title: Text('Remove "${existing.name}"?'),
        content: const Text('This only removes the checklist, not your log history.'),
        actions: [
          TextButton(onPressed: () => Navigator.pop(ctx, false), child: const Text('Cancel')),
          TextButton(onPressed: () => Navigator.pop(ctx, true), child: Text('Remove', style: TextStyle(color: AppColors.of(ctx).alert))),
        ],
      ),
    );
    if (confirmed == true) {
      widget.onUpdateVehicle(removeService(widget.vehicle, existing.id));
      widget.onBack();
    }
  }

  @override
  Widget build(BuildContext context) {
    final c = AppColors.of(context);
    final categories = widget.vehicle.categories;

    return Container(
      color: c.bg,
      child: Column(
        children: [
          TopBar(title: _existing != null ? 'Edit service' : 'New service', onBack: widget.onBack),
          Expanded(
            child: ListView(
              padding: const EdgeInsets.fromLTRB(AppSpace.side, 12, AppSpace.side, 40),
              keyboardDismissBehavior: ScrollViewKeyboardDismissBehavior.onDrag,
              children: [
                AppCard(children: [
                  AppField(label: 'Name', value: _name, placeholder: 'e.g. Chain weekend', last: true, onChanged: (v) => _name = v),
                ]),
                const SizedBox(height: AppSpace.block),
                const Label('Items', margin: EdgeInsets.only(bottom: 10)),
                if (categories.isEmpty)
                  Text('Add categories first, then pick them here.', style: AppTypography.small.copyWith(color: c.muted))
                else
                  AppCard(children: [
                    for (var i = 0; i < categories.length; i++)
                      CheckRow(
                        label: categories[i].name,
                        checked: _selected.contains(categories[i].id),
                        onToggle: () => _toggle(categories[i].id),
                        last: i == categories.length - 1,
                      ),
                  ]),
                ActionRow(label: 'Save service', onPressed: _save),
                if (_existing != null) ActionRow(label: 'Remove service', onPressed: _remove, destructive: true),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

File: src/screens/EditServiceScreen.js
import React, { useState } from "react";
import { View, Text, ScrollView, Alert } from "react-native";
import { Card, Label, Field, Action, CheckRow, useC } from "../components/ui";
import { TYPE, SPACE } from "../theme";
import { uid, getServiceRecords, upsertService, removeService } from "../data/templates";
import TopBar from "../components/TopBar";

export default function EditServiceScreen({ vehicle, serviceId, onBack, onUpdateVehicle }) {
  const c = useC();
  const existing = serviceId ? getServiceRecords(vehicle).find((s) => s.id === serviceId) : null;
  const [name, setName] = useState(existing?.name || "");
  const [selected, setSelected] = useState(() => new Set(existing?.categoryIds || []));

  const toggle = (id) => {
    setSelected((prev) => {
      const next = new Set(prev);
      if (next.has(id)) next.delete(id);
      else next.add(id);
      return next;
    });
  };

  const save = () => {
    const trimmed = name.trim();
    if (!trimmed) return;
    const categoryIds = vehicle.categories.map((cat) => cat.id).filter((id) => selected.has(id));
    if (!categoryIds.length) return;
    onUpdateVehicle(
      upsertService(vehicle, {
        id: existing?.id || uid(),
        name: trimmed,
        categoryIds,
      })
    );
    onBack();
  };

  const remove = () => {
    Alert.alert(`Remove "${existing.name}"?`, "This only removes the checklist, not your log history.", [
      { text: "Cancel", style: "cancel" },
      {
        text: "Remove",
        style: "destructive",
        onPress: () => {
          onUpdateVehicle(removeService(vehicle, existing.id));
          onBack();
        },
      },
    ]);
  };

  return (
    <View style={{ flex: 1, backgroundColor: c.bg }}>
      <TopBar title={existing ? "Edit service" : "New service"} onBack={onBack} />
      <ScrollView
        contentContainerStyle={{ paddingHorizontal: SPACE.side, paddingTop: 12, paddingBottom: 40 }}
        keyboardShouldPersistTaps="handled"
      >
        <Card>
          <Field
            label="Name"
            value={name}
            onChangeText={setName}
            placeholder="e.g. Chain weekend"
            last
          />
        </Card>

        <View style={{ marginTop: SPACE.block }}>
          <Label style={{ marginBottom: 10 }}>Items</Label>
          {vehicle.categories.length === 0 ? (
            <Text style={[TYPE.small, { color: c.muted }]}>Add categories first, then pick them here.</Text>
          ) : (
            <Card>
              {vehicle.categories.map((cat, i) => (
                <CheckRow
                  key={cat.id}
                  label={cat.name}
                  checked={selected.has(cat.id)}
                  onToggle={() => toggle(cat.id)}
                  last={i === vehicle.categories.length - 1}
                />
              ))}
            </Card>
          )}
        </View>

        <Action label="Save service" onPress={save} />
        {existing ? <Action label="Remove service" onPress={remove} destructive /> : null}
      </ScrollView>
    </View>
  );
}
